logging/src/logging/filter.cpp

```diff
@@ -0,0 +1,507 @@
+               ? fr::accept
+               : fr::reject;
+    }
+
+    virtual
+    dynamic_t
+    representation() const {
+        return dynamic_t(std::make_tuple(std::string("e"), attribute_name));
+    };
+
+    std::string attribute_name;
+};
+
+struct not_exists_filter_t: public filter_t::filter_impl_t {
+    not_exists_filter_t(std::string _attribute_name) :
+        attribute_name(std::move(_attribute_name))
+    {}
+
+    virtual
+    filter_result_t
+    apply(const std::string&, unsigned int, const logging::attributes_t& attributes) const {
+        return static_cast<bool>(logging::find_attribute(attributes, attribute_name))
+               ? fr::reject
+               : fr::accept;
+    }
+
+    virtual
+    dynamic_t
+    representation() const {
+        return dynamic_t(std::make_tuple(std::string("!e"), attribute_name));
+    };
+
+    std::string attribute_name;
+};
+
+template<class T>
+struct comparision_filter_t: public filter_t::filter_impl_t {
+    comparision_filter_t(std::string _attribute_name, T _attribute_value):
+        attribute_name(std::move(_attribute_name)),
+        attribute_value(std::move(_attribute_value))
+    {}
+
+    std::string attribute_name;
+    T attribute_value;
+};
+
+template<class T>
+struct equals_filter_t: public comparision_filter_t<T> {
+    equals_filter_t(std::string _attribute_name, T _attribute_value):
+        comparision_filter_t<T>(std::move(_attribute_name), std::move(_attribute_value))
+    {}
+
+    virtual
+    filter_result_t
+    apply(const std::string&, unsigned int, const logging::attributes_t& attributes) const {
+        const auto& val = logging::find_attribute(attributes, this->attribute_name);
+        return (val && val->value.template convertible_to<T>() && val->value.template to<T>() == this->attribute_value)
+                ? fr::accept
+                : fr::reject;
+    }
+
+    virtual
+    dynamic_t
+    representation() const {
+        return dynamic_t(std::make_tuple(std::string("=="), this->attribute_name, this->attribute_value));
+    };
+};
+
+template<class T>
+struct not_equals_filter_t: public comparision_filter_t<T> {
+    not_equals_filter_t(std::string _attribute_name, T _attribute_value):
+        comparision_filter_t<T>(std::move(_attribute_name), std::move(_attribute_value))
+    {}
+
+    virtual
+    filter_result_t
+    apply(const std::string&, unsigned int, const logging::attributes_t& attributes) const {
+        const auto& val = logging::find_attribute(attributes, this->attribute_name);
+        return (val && val->value.template convertible_to<T>() && val->value.template to<T>() == this->attribute_value)
+                ? fr::reject
+                : fr::accept;
+    }
+
+    virtual
+    dynamic_t
+    representation() const {
+        return dynamic_t(std::make_tuple(std::string("!="), this->attribute_name, this->attribute_value));
+    };
+};
+
+template<class T>
+struct greater_filter_t: public comparision_filter_t<T> {
```

Comment:
Ok, these filters API looks universal and powerful, but let me be Vanga - noone will never ever use it.